### DISCUSSION TITLE
Solver don't keep the format of the DensityMatrix

### DISCUSSION THREAD

**Discussion by ArturDomingues** [qiskit-community/qiskit-dynamics#95] [Q&A] [answered]
I was trying to use partial trace in the density matrices after using the solver.solve method, but this method "blends" the format of my density matrices. What do I mean with "blend" is that I start with a Zero.tensor(I), which in my problem Zero is [[1,0,0],[0,0,0],[0,0,0]] and I is the 2x2 identity, and end up with a 6x6 matrix, which is ok, but the format starts as (3,2) dimension density matrix and end up with a (6,) dimension density matrix, I really don`t know what to do, since I want to do partial trace out the 2x2 subsystem in my analyses. (Sorry about all the questions in these last months)

**Comment by DanPuzzuoli** [comment]
Ah interesting - thanks for pointing this out. Nothing to apologize for! It's great to have users who point out issues like this :) .

This is definitely a problem - Dynamics respects the `quantum_info` states, but doesn't preserve the metadata, like subsystem dimensions. Basically, `solve` checks what type the input state is to determine what to do, pulls out the internal array to actually do the computation, then rewraps the output in the right state type before returning. Unfortunately it doesn't keep track of the subsystem dimensions, so when rewrapping it just gets set to the default (which is just the dimension).

I'll make an issue for this as something that needs to be fixed in Dynamics. For now though you'll need to manually reset the subsystem dimensions in the `DensityMatrix`, e.g. if your density matrix is `yf`:
```
yf = DensityMatrix(yf.data, dims=(3, 2))
```

Doing `yf.data` pulls out the internal array, which you can then rewrap in `DensityMatrix` with the right dimensions.

**Reply by ArturDomingues** [reply]
It didn't work using yf.data, but I used 
```
for t, sol in enumerate(yf.y):
  yf.y[t] = DensityMatrix(sol, dims = (3, 2))
```
I didn't test if this is the proper DensityMatrix output, but It raise an error when I tried to use 
`yf= DensityMatrix(yf.data, dims=(3, 2))`

**Reply by DanPuzzuoli** [reply]
Ah yes sorry, I was taking, e.g.:
```
results = solver.solve(...)

yf = results.y[-1]
```
As you did in your solution, you need to correct the density matrix for every entry of `results.y`.

**Comment by DanPuzzuoli** [comment, marked as answer]
@ArturDomingues just FYI PR #99 fixes this problem. We are hoping to have a release in the next few weeks so this issue should be resolved soon!
